File: ios/AdaptiveTextTokenizer.swift
import Foundation

/// A single token in the rendered flow. Tokens have stable identities so
/// SwiftUI knows which subview moved between layout passes (which is what
/// makes individual words animate to new line positions).
struct AdaptiveTextToken: Identifiable, Equatable {
  let id: String
  let text: String
  /// Whether this token must remain attached to the previous one (no
  /// inter-token spacing). Used so trailing punctuation like `.` or `,`
  /// doesn't drift away from its word when the line wraps.
  let attachToPrevious: Bool
}

enum AdaptiveTextTokenizer {
  /// Splits `text` into tokens based on `mode`. The returned IDs are stable
  /// within a tokenization pass: identical strings at the same position
  /// produce the same ID, so SwiftUI treats them as the same logical view
  /// across renders and animates frame deltas.
  static func tokenize(_ text: String, mode: AdaptiveTextSplitMode) -> [AdaptiveTextToken] {
    switch mode {
    case .word:
      return tokenizeWords(text)
    case .grapheme:
      return tokenizeGraphemes(text)
    }
  }

  private static func tokenizeWords(_ text: String) -> [AdaptiveTextToken] {
    guard !text.isEmpty else { return [] }
    var tokens: [AdaptiveTextToken] = []
    var indexCounter = 0
    let punctuation = CharacterSet(charactersIn: ",.!?;:)]}\"'")

    // Split on Unicode whitespace; preserve internal punctuation as part of
    // each token so words don't fragment ("hello," stays one token).
    let pieces = text.components(separatedBy: .whitespacesAndNewlines)
    for piece in pieces where !piece.isEmpty {
      // Check whether this piece is _only_ punctuation that should attach
      // to the prior word (rare with whitespace splitting, but possible if
      // someone wrote "hello ,world").
      let attach = piece.unicodeScalars.allSatisfy { punctuation.contains($0) }
      tokens.append(
        AdaptiveTextToken(
          id: "w\(indexCounter):\(piece)",
          text: piece,
          attachToPrevious: attach && !tokens.isEmpty
        )
      )
      indexCounter += 1
    }
    return tokens
  }

  private static func tokenizeGraphemes(_ text: String) -> [AdaptiveTextToken] {
    var tokens: [AdaptiveTextToken] = []
    var i = 0
    for cluster in text {
      let s = String(cluster)
      // Skip pure whitespace graphemes — they collapse into the inter-token
      // gap so the layout stays clean.
      if s.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
        i += 1
        continue
      }
      tokens.append(
        AdaptiveTextToken(id: "g\(i):\(s)", text: s, attachToPrevious: false)
      )
      i += 1
    }
    return tokens
  }
}

File: android/src/main/java/com/adaptivetext/compose/AdaptiveTextTokenizer.kt
package com.adaptivetext.compose

import java.text.BreakIterator
import java.util.Locale

/**
 * A single token in the rendered flow. Stable [id] strings let Compose
 * smart-skip and animate position changes — words that survive a re-layout
 * keep their identity and `Modifier.animateBounds` interpolates them.
 */
data class AdaptiveTextToken(
  val id: String,
  val text: String,
  val attachToPrevious: Boolean,
)

object AdaptiveTextTokenizer {
  private val PUNCTUATION = setOf(',', '.', '!', '?', ';', ':', ')', ']', '}', '\'', '"')

  fun tokenize(text: String, mode: SplitMode): List<AdaptiveTextToken> = when (mode) {
    SplitMode.WORD -> tokenizeWords(text)
    SplitMode.GRAPHEME -> tokenizeGraphemes(text)
  }

  private fun tokenizeWords(text: String): List<AdaptiveTextToken> {
    if (text.isEmpty()) return emptyList()
    val pieces = text.split(Regex("\\s+")).filter { it.isNotEmpty() }
    return pieces.mapIndexed { index, piece ->
      val attach = piece.all { it in PUNCTUATION }
      AdaptiveTextToken(
        id = "w$index:$piece",
        text = piece,
        attachToPrevious = attach && index > 0,
      )
    }
  }

  private fun tokenizeGraphemes(text: String): List<AdaptiveTextToken> {
    val out = mutableListOf<AdaptiveTextToken>()
    val iter = BreakIterator.getCharacterInstance(Locale.getDefault())
    iter.setText(text)
    var start = iter.first()
    var idx = 0
    var end = iter.next()
    while (end != BreakIterator.DONE) {
      val cluster = text.substring(start, end)
      if (cluster.isNotBlank()) {
        out.add(AdaptiveTextToken(id = "g$idx:$cluster", text = cluster, attachToPrevious = false))
      }
      start = end
      end = iter.next()
      idx++
    }
    return out
  }
}
